Src/WitsmlExplorer.Api/Services/CredentialsService.cs

```diff
@@ -108,39 +92,127 @@ private string Decrypt(string inputString)
+                {
+                    Host = new Uri(headers.GetHeaderValue(server)),
+                    UserId = cacheContents.Split(":")[0],
+                    Password = cacheContents.Split(":")[1]
+                } :
+                null;
+        }
 
+        private static string GetClaimFromTokenValue(string token, string claim)
+        {
             JwtSecurityTokenHandler handler = new();
-            JwtSecurityToken jwt = handler.ReadJwtToken(token.Split()[1]);
-            return jwt.Claims.Where(n => n.Type == "upn").Select(n => n.Value).FirstOrDefault();
+            JwtSecurityToken jwt = handler.ReadJwtToken(token);
+            return jwt.Claims.Where(n => n.Type == claim).Select(n => n.Value).FirstOrDefault();
+        }
+
+        public string GetClaimFromToken(IEssentialHeaders headers, string claim)
+        {
+            return GetClaimFromTokenValue(headers.GetBearerToken(), claim);
         }
 
         private ServerCredentials GetBasicCredentialsFromHeaderValue(string headerValue)
         {
             return HttpRequestExtensions.ParseServerHttpHeader(headerValue, Decrypt);
         }
 
-        public (string userPrincipalName, string witsmlUserName) GetUsernamesFromCookieAndToken(EssentialHeaders headers)
+        public (ServerCredentials targetServer, ServerCredentials sourceServer) GetWitsmlUsernamesFromCache(IEssentialHeaders headers)
+        {
+            (ServerCredentials witsmlTargetCredentials, ServerCredentials witsmlSourceCredentials) result;
+            if (headers.GetCookieValue() != null)
+            {
+                ServerCredentials target = GetWitsmlCredentialsByCookie(headers, EssentialHeaders.WitsmlTargetServer).Result;
+                ServerCredentials source = GetWitsmlCredentialsByCookie(headers, EssentialHeaders.WitsmlSourceServer).Result;
+                result = (target, source);
+            }
+            else
+            {
+                ServerCredentials target = GetWitsmlCredentialsByToken(headers, EssentialHeaders.WitsmlTargetServer).Result;
+                ServerCredentials source = GetWitsmlCredentialsByToken(headers, EssentialHeaders.WitsmlSourceServer).Result;
+                result = (target, source);
+            }
+            return result;
+        }
+        private Task<ServerCredentials> GetWitsmlCredentialsByCookie(IEssentialHeaders headers, string server)
         {
-            ServerCredentials witsmlCredentials = GetCredentialsCookieFirst(headers, EssentialHeaders.WitsmlTargetServer).Result;
-            return (GetTokenUserPrincipalName(headers.Authorization), witsmlCredentials.UserId);
+            ServerCredentials result;
+            if (string.IsNullOrEmpty(headers.GetCookieValue()) || string.IsNullOrEmpty(headers.GetHeaderValue(server)))
+            {
+                result = new ServerCredentials();
+            }
+            else
+            {
+                string host = new Uri(headers.GetHeaderValue(server)).Host;
```

Comment:
Doesn't this bit pretty much duplicate GetCredentialsFromCache?

Reply:
I think you are correct. They probably started out on different places and have now the same logic. `GetCredentialsFromCache` can be used to fulfill `GetWitsmlCredentialsByCookie`. I will verify and remove if so.

Reply:
Was able to remove both `GetWitsmlCredentialsByCookie` and `GetWitsmlCredentialsByToken` now so thanks! :)